reporter/src/main/kotlin/reporters/NoticeByPackageReporter.kt

```diff
@@ -0,0 +1,270 @@
+            }
+
+            addLicenseFileFindings(licenseFileFindings, archiveDir)
+
+            val licensesInNoticeFiles = licenseFileFindings.values.map { it.keys }.flatten().toSet()
+            val processedFindings = licenseFindingsMap.removeGarbage(input.copyrightGarbage).processStatements()
+                .filter { (license, _) -> license !in licensesInNoticeFiles }
+                .filter { (license, _) ->
+                    input.licenseTextProvider.hasLicenseText(license).also {
+                        if (!it) {
+                            log.warn {
+                                "No license text found for license '$license', it will be omitted from the report."
+                            }
+                        }
+                    }
+                }.toSortedMap()
+
+            if (processedFindings.isNotEmpty()) {
+                if (licenseFileFindings.isNotEmpty()) {
+                    add { LICENSE_SEPARATOR }
+                }
+
+                add { "The following copyrights and licenses were found in the source code of this package:\n\n" }
+            }
+
+            addProcessedFindings(processedFindings)
+
+            if (licenseFileFindings.isEmpty() && processedFindings.isEmpty()) {
+                log.error { "No license information was added for package ${id.toCoordinates()}." }
+            }
+        }
+    }
+
+    private fun MutableList<() -> String>.addProcessedFindings(processedFindings: LicenseFindingsMap) {
+        processedFindings.entries.forEachIndexed { index, (license, copyrights) ->
+            input.licenseTextProvider.getLicenseText(license)?.let { licenseText ->
+                copyrights.forEach { copyright ->
+                    add { "$copyright\n" }
+                }
+                if (copyrights.isNotEmpty()) {
+                    add { "\n" }
+                }
+
+                add { licenseText }
+
+                if (index < processedFindings.size - 1) {
+                    add { LICENSE_SEPARATOR }
+                }
+            }
+        }
+    }
+
+    private fun MutableList<() -> String>.addLicenseFileFindings(
+        licenseFileFindings: Map<String, LicenseFindingsMap>,
+        archiveDir: File
+    ) {
+        licenseFileFindings.forEach { (file, findings) ->
+            add { "This package contains the file $file with the following contents:\n\n" }
+            add { "${archiveDir.resolve(file).readText()}\n" }
+            val allCopyrights = findings.values.flatten().toSet()
+            val processedCopyrights = CopyrightStatementsProcessor().process(allCopyrights).toMutableSet()
+            if (processedCopyrights.isNotEmpty()) {
+                add { "The following copyright holder information relates to the license(s) above:\n\n" }
+            }
+            processedCopyrights.forEach { copyright ->
+                add { "$copyright\n" }
+            }
+        }
+    }
+
+    private fun getFindingsForLicenseFiles(
```

Comment:
Could you please describe the idea of the algorithm somewhere?

Reply:
I have added a short description to the function, maybe that helps?